### PR TITLE
Backend groups

### PR DESCRIPTION
**Pull request by mtte**
Implemented all group endpoints according to the API spec. The list user had to be implemented too as it was needed by a group endpoint.

### CODE REVIEW

File: api/src/routes.php [hunk 1]
```diff
@@ -17,33 +19,65 @@
...
     $this->get('/events', EventsController::class . ':listEvents')
         ->add(new ValidationErrorResponseBuilder())
-        ->add(new Validation($this->getContainer()['validation']['listEvents']))
-        ->add(new AuthenticationMiddleware($container, \JMP\Utils\PermissionLevel::USER))
+        ->add(new Validation($container['validation']['listEvents']))
+        ->add(new AuthenticationMiddleware($container, PermissionLevel::USER))
         ->add($jwtMiddleware);
 
-    $this->get('/events/{id}', EventsController::class . ':getEventById')
+    $this->get('/events/{id:[0-9]+}', EventsController::class . ':getEventById')
```
Comment (dominikstraessle) [inline api/src/routes.php]: Is this regex useful or should we move all validation into validation.php? Update all routes with a such regex, where possible, e.g. /registration/eventId/userId route

File: api/src/JMP/Controllers/GroupsController.php [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        if ($group->isFailure()) {
+            return $this->groupIdNotAvailable($response, $id);
+        }
+        $group = $group->getData();
+
+
+        $newName = $request->getParsedBodyParam('name');
+
+        // check if a new name is specified and is different from the current name
+        if (is_null($newName) || $newName === $group->name) {
```
[the remark this answers is not in the document]
Comment (mtte) [reply api/src/JMP/Controllers/GroupsController.php]: No as the parameter for the update route is optional. Therefore the param "name" can be `null`.

File: api/src/JMP/Services/GroupService.php [hunk 3]
```diff
@@ -13,37 +13,206 @@ class GroupService
...
+
+        return $this->getGroupByName($name);
+    }
+
+    /**
+     * @param $id
+     */
+    public function deleteGroup($id): void
+    {
+        $this->deleteMemberships($id);
```
[the remark this answers is not in the document]
Comment (mtte) [reply api/src/JMP/Services/GroupService.php]: I'd rather not do that as you could more easily delete data by accident. In this case the foreign keys are deleted explicitly. Are you ok with that?

File: api/src/JMP/Services/GroupService.php [hunk 4]
```diff
@@ -13,37 +13,206 @@ class GroupService
...
+        $sql = <<< SQL
+        SELECT id, name
+        FROM `group`
+        WHERE name = :name
+SQL;
+
+        $stmt = $this->db->prepare($sql);
+        $stmt->bindParam(':name', $name);
+        $stmt->execute();
+
```
[the remark this answers is not in the document]
Comment (mtte) [reply api/src/JMP/Services/GroupService.php]: This should habe been a private method that is only used by the group service itself when its clear that the group name is existing. I added a check with `if($stmt->rowCount() < 1)` that throws an exception if that is the case.

File: api/src/JMP/Services/GroupService.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (dominikstraessle) [reply api/src/JMP/Services/GroupService.php]: 👍

**Pull request merged**